use inferred type instead of `Union{}` for empty comprehensions

JeffBezanson opened the pull request:
This implements the behavior requested in #17811. To many people, it is probably the expected behavior for 0.5.

JeffBezanson committed:
use inferred type instead of `Union{}` for empty comprehensions

fixes #17811

iamed2 commented:
Is this going to be merged into 0.5 too then?

tkelman commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

also maybe worth comparing vs rc0, or would this not make much of a difference?

JeffBezanson commented:
Comparing to master is probably good enough.

@iamed2 Yes I think this is a candidate for inclusion in 0.5.

tkelman commented:
@jrevels how do we get the full error there?

tkelman commented:
feels like this should have some tests

jrevels commented:
Nanosoldier is down at the moment while I roll it over to v0.6, sorry about that.

jrevels commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17847#issuecomment-238420851) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2c2b99d99b7ffb7cf8173069cf119a657bfd9315/87999ad_vs_2037464/report.md). cc @jrevels

ranjanan committed:
test for #17811, type of e.g. `abs(Integer[])` (#17891)

This tests that the result type of an empty comprehension is big enough to hold the possible results.

johnmyleswhite commented:
For 0.6 we should decide whether this empty case logic also applies to operations on `Nullable`. Cf. #16961

stevengj commented:
This logic should also apply to `map` and `broadcast`.
